### PR TITLE
Allow optional block number parameter

### PR DESCRIPTION
**Pull request by nebojsa94**
This PR allows optional block number parameter in RPC methods eth_call, eth_estimateGas, eth_getBalance, eth_getTransactionsCount, eth_getStorageAt and eth_getCode.

### CODE REVIEW

**Comment by alcuadrado** [comment]
Hey @nebojsa94,

Thanks for submitting this PR. As Andrej probably told you, we'd like to release this soon, so you can start testing your integration with it.

Can I request two simple changes? I'm worried that how this is built rn will make the integration with #768 harder.

These are:

1. Instead of getting the `storageRoot` in the eth module, pass the block number to the node and manage that there. The reason for this is that when forking mainnet you have to forward some requests to a node through JSON-RPC and there's no direct way to do that using the storage root as a param.

2. Can we introduce some kind of helper into the node so the set/reset storage root isn't repeated? Something like this should do:

```ts
private async _runOnBlockContext<T>(blockNumber:BN, action:() => Promise<T>):Promise<T> {
     // get storage root by block number
     // set the storage root
    try {
       return await action(); // I think the await here is crucial
    } finally {
         // reset the storage root
    }
}
```

**Comment by fvictorio** [comment]
Yep, that makes sense!

File: packages/buidler-core/test/internal/buidler-evm/provider/modules/eth.ts [hunk 1]
```diff
@@ -421,6 +463,43 @@ describe("Eth module", function () {
...
+
+          const result2 = await this.provider.send("eth_estimateGas", [
+            {
+              to: contractAddress,
+              from: DEFAULT_ACCOUNTS_ADDRESSES[0],
+              data: EXAMPLE_CONTRACT.selectors.modifiesState + newState,
+            },
+          ]);
+
+          assert.isTrue(new BN(toBuffer(result)).gt(new BN(toBuffer(result2))));
```
Comment (alcuadrado) [inline packages/buidler-core/test/internal/buidler-evm/provider/modules/eth.ts:501]: clever!

File: packages/buidler-core/src/internal/buidler-evm/provider/modules/eth.ts [hunk 2]
```diff
@@ -1036,32 +1041,31 @@ export class EthModule {
...
 
-  private async _validateBlockTag(blockTag: OptionalBlockTag) {
-    const latestBlock = await this._node.getLatestBlockNumber();
-
-    if (BN.isBN(blockTag) && latestBlock.eq(blockTag)) {
-      return;
-    }
+  private async _blockTagToBlockNumber(
+    blockTag: OptionalBlockTag
+  ): Promise<BN> {
```
Comment (alcuadrado) [inline packages/buidler-core/src/internal/buidler-evm/provider/modules/eth.ts:1050]: Make this `BN | null`.

File: packages/buidler-core/src/internal/buidler-evm/provider/modules/eth.ts [hunk 3]
```diff
@@ -1036,32 +1041,31 @@ export class EthModule {
...
 
-    // We only support latest and pending. As this provider doesn't have pending transactions, its
-    // actually just latest.
     if (
-      blockTag !== undefined &&
-      blockTag !== "latest" &&
-      blockTag !== "pending"
+      blockTag === undefined ||
+      blockTag === "latest" ||
+      blockTag === "pending"
```
Comment (alcuadrado) [inline packages/buidler-core/src/internal/buidler-evm/provider/modules/eth.ts:1056]: If `"pending"`, return `null`.

File: packages/buidler-core/src/internal/buidler-evm/provider/node.ts [hunk 4]
```diff
@@ -1437,6 +1470,27 @@ If you are using a wallet or dapp, try resetting your wallet's accounts.`
     return parentTd.add(difficulty);
   }
 
+  private async _runOnBlockContext<T>(
+    blockNumber: BN | null,
```
Comment (alcuadrado) [inline packages/buidler-core/src/internal/buidler-evm/provider/node.ts:1474]: `null` here would mean `pending`, which for now is the latest block, so no modification is needed.

File: packages/buidler-core/src/internal/buidler-evm/provider/node.ts [hunk 5]
```diff
@@ -491,7 +491,8 @@ export class BuidlerNode extends EventEmitter {
 
   public async runCall(
     call: CallParams,
-    runOnNewBlock: boolean
+    runOnNewBlock: boolean,
```
Comment (alcuadrado) [inline packages/buidler-core/src/internal/buidler-evm/provider/node.ts:494]: `runOnNewBlock` can go away

File: packages/buidler-core/src/internal/buidler-evm/provider/node.ts [hunk 6]
```diff
@@ -500,10 +501,12 @@ export class BuidlerNode extends EventEmitter {
   }> {
     const tx = await this._getFakeTransaction({
       ...call,
-      nonce: await this.getAccountNonce(call.from),
+      nonce: await this.getAccountNonce(call.from, null),
     });
 
-    const result = await this._runTxAndRevertMutations(tx, runOnNewBlock);
+    const result = await this._runOnBlockContext(blockNumber, () =>
+      this._runTxAndRevertMutations(tx, runOnNewBlock)
```
Comment (alcuadrado) [inline packages/buidler-core/src/internal/buidler-evm/provider/node.ts:508]: Here `runOnNewBlock` should be `blockNumber === null`.

**Review by alcuadrado — CHANGES_REQUESTED** [review]
Sorry, I didn't notice this before, but it's a bit confusing how block numbers are handled when using "pending", especially in `runCall`.

I think we can use `null` to represent "pending", and BNs everywhere else.

**Comment by alcuadrado** [comment]
I have push access to your repo, so I fixed those things. Thanks for doing the changes I asked for @nebojsa94

**Pull request merged**